src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicBuffers.cs

```diff
@@ -0,0 +1,90 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using Microsoft.Quic;
+
+namespace System.Net.Quic.Implementations.MsQuic.Internal;
+
+/// <summary>
+/// Helper class to convert managed data into QUIC_BUFFER* consumable by MsQuic.
+/// It also allows reuse with repeated Reset/Initialize calls.
+/// Note that since this is struct and there's no finalizer, Dispose must be always called to release the unmanaged memory allocated by this struct.
+/// </summary>
+internal unsafe struct MsQuicBuffers : IDisposable
+{
+    // Handles to pinned memory blocks from the user.
+    private MemoryHandle[] _handles;
+    // Native memory block which holds the pinned memory pointers from _handles and can be passed to MsQuic as QUIC_BUFFER*.
+    private IntPtr _buffers;
+    // Number of QUIC_BUFFER instance currently allocated in _buffers, so that we can reuse the memory instead of reallocating.
+    private int _count;
+
+    public MsQuicBuffers()
+    {
+        _handles = Array.Empty<MemoryHandle>();
+        _buffers = IntPtr.Zero;
+        _count = 0;
+    }
+
+    public QUIC_BUFFER* Buffers => (QUIC_BUFFER*)_buffers;
+    public int Count => _count;
+
+    /// <summary>
+    /// The mothod initializes QUIC_BUFFER* with data from inputs, converted via toBuffer.
+    /// Note that the struct either needs to be freshly created or previously cleaned up with Reset.
+    /// </summary>
+    /// <param name="inputs">Inputs to get their byte array, pin them and pepare them to be passed to MsQuic as QUIC_BUFFER*.</param>
+    /// <param name="toBuffer">Method extracting byte array from the inputs, e.g. applicationProtocol.Protocol.</param>
+    /// <typeparam name="T">The type of the inputs.</typeparam>
+    public void Initialize<T>(IList<T> inputs, Func<T, ReadOnlyMemory<byte>> toBuffer)
+    {
+        if (_handles.Length < inputs.Count)
+        {
+            ArrayPool<MemoryHandle>.Shared.Return(_handles);
+            _handles = ArrayPool<MemoryHandle>.Shared.Rent(inputs.Count);
+        }
+        if (_count < inputs.Count)
+        {
+            Marshal.FreeHGlobal(_buffers);
+            _buffers = Marshal.AllocHGlobal(sizeof(QUIC_BUFFER) * inputs.Count);
+        }
+
+        _count = inputs.Count;
+
+        var buffers = Buffers;
+        for (int i = 0; i < inputs.Count; ++i)
+        {
+            var buffer = toBuffer(inputs[i]);
+            var handle = buffer.Pin();
+
+            _handles[i] = handle;
+            buffers[i].Buffer = (byte*)handle.Pointer;
+            buffers[i].Length = (uint)buffer.Length;
+        }
+    }
+
+    /// <summary>
+    /// Unpins the managed memory and allows reuse of this struct.
```

Comment:
It may be best to add this only once the callers are actually using this. 

As implemented, the contract of this method is very odd (it leaves the struct in half-disposed half-initialized state). This implementation will very likely require changes for reliability to allow reuse of this struct.

Reply:
That's kind of the point, but I might be doing something wrong. `Reset` here is based on what we do with send buffers in `MsQuicStream`:
https://source.dot.net/#System.Net.Quic/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs,fe7e534486d47fbe,references

Since I planned to reuse this there as well, just not in this PR.

So if you see a problem with it, we might already have an issue in our current code and I'd appreciate some guidance here.